libsigner/src/runloop.rs

```diff
@@ -0,0 +1,244 @@
+// Copyright (C) 2013-2020 Blockstack PBC, a public benefit corporation
+// Copyright (C) 2020-2023 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#![allow(unused_imports)]
+#![allow(dead_code)]
+
+use std::marker::PhantomData;
+use std::net::SocketAddr;
+use std::sync::mpsc::{channel, Receiver, RecvTimeoutError};
+use std::thread;
+use std::thread::JoinHandle;
+use std::time::Duration;
+
+use crate::events::{EventReceiver, EventStopSignaler, StackerDBChunksEvent};
+
+use crate::error::EventError;
+
+use stacks_common::deps_common::ctrlc as termination;
+use stacks_common::deps_common::ctrlc::SignalId;
+
+use libc;
+
+/// Some libcs, like musl, have a very small stack size.
+/// Make sure it's big enough.
+const THREAD_STACK_SIZE: usize = 128 * 1024 * 1024; // 128 MB
+
+/// stderr fileno
+const STDERR: i32 = 2;
+
+/// Trait describing the needful components of a top-level runloop.
+/// This is where the signer business logic would go.
+/// Implement this, and you get all the multithreaded setup for free.
+pub trait SignerRunLoop<R> {
+    /// Hint to set how long to wait for new events
+    fn set_event_timeout(&mut self, timeout: Duration);
+    /// Getter for the event poll timeout
+    fn get_event_timeout(&self) -> Duration;
+    /// Run one pass of the event loop, given new StackerDB events discovered since the last pass.
+    /// Returns Some(R) if this is the final pass -- the runloop evaluated to R
+    /// Returns None to keep running.
+    fn run_one_pass(&mut self, events: Option<StackerDBChunksEvent>) -> Option<R>;
```

Comment:
"events" suggests there are more than one event. Should this be a collection/iterator instead of an option?